src/main/java/org/walkmod/javalang/compiler/types/IndexedURLClassPath.java

```
@@ -0,0 +1,159 @@
+            if (lastIndexed < urls.length) {
+                indexURLs(urls[lastIndexed]);
+                lastIndexed ++;
+                return getResource(name, check);
+            }
+            return null;
+        }
+        return delegate.getResource(name, check);
+    }
+
+    @Override
+    public URL findResource(final String name, boolean check) {
+        URLClassPath delegate = index.get(name);
+        if (delegate == null) {
+            if (lastIndexed < urls.length) {
+                indexURLs(urls[lastIndexed]);
+                lastIndexed ++;
+                return findResource(name, check);
+            }
+            return null;
+        }
+        return delegate.findResource(name, check);
+    }
+
+
+    public List<String> listPackageContents(final String packageName, boolean check) {
+
+        String packageFile = packageName.replaceAll("\\.", File.separator);
+        while (lastIndexed < urls.length) {
+            indexURLs(urls[lastIndexed]);
+            lastIndexed ++;
+        }
+        return index.list(packageFile);
+    }
+
+    public List<String> listSDKContents(final String packageName, boolean check) {
+        String packageFile = packageName.replaceAll("\\.", File.separator);
+        indexURLs(RT_JAR);
+        return index.list(packageFile);
+    }
+
+
+    private void indexURLs(URL url) {
+       try {
+           if (!"file".equals(url.getProtocol())) {
+               throw new RuntimeException("Classpath element is not a file: " + url);
+           }
+           File root = new File(url.getPath());
+           URL[] args = {url};
+           URLClassPath delegate = new URLClassPath(args);
+
+           if (root.isDirectory()) {
+               String rootPath = root.getPath();
+               if (!rootPath.endsWith(File.separator)) {
+                   rootPath = rootPath + File.separator;
+               }
+               addFilesToIndex(rootPath.length(), root, delegate);
```

Whats the urlclasspath definition for file archive?
In the past we had .zip accepted. Years ago but would check against the urlclassloader.

---

URLClasspath belongs to the "misc" classes of the JVM to load resources.
I have excluded zips because I did not about this, but I can include them as well.

---

I suggest checking how urlclassloader handles non-directories and do the same. If they check for suffixes we can do the same. But I expectvsomething different.

---

URLConnection urlc = url.openConnection();
            InputStream is = urlc.getInputStream();
            if (urlc instanceof JarURLConnection) {

is what urclassloader does.
I suggest trying to open each non-directory as jar and ignore errors if it's not a jar.

---

This is not right at all. The `URLClassLoader` is checking the permissions for the file with the approach that you pointed out. 
http://grepcode.com/file/repository.grepcode.com/java/root/jdk/openjdk/6-b14/java/net/URLClassLoader.java#458

However, here: http://grepcode.com/file/repository.grepcode.com/java/root/jdk/openjdk/6-b14/sun/misc/URLClassPath.java#URLClassPath.JarLoader

---

I don't see your point/contradiction.

---

In the JarLoader is not opening a connection, it is iterating the entries. Looking in detail in the code, I am starting to think that I should maintain and index in `TypesLoaderVisitor` but using the current `URLClassLoader`, which already is indexing the entries.

---

Nevermind. It not indexes the classpath at the begining. Every request is really expensive in the `URLClassLoader` class